File: modules/vietride-places/ios/VietRidePlacesModule.swift
import ExpoModulesCore
import Foundation

#if canImport(GooglePlaces)
import GooglePlaces
import CoreLocation
#endif

public class VietRidePlacesModule: Module {
  #if canImport(GooglePlaces)
  private let sessionLock = NSLock()
  private var sessions: [String: GMSAutocompleteSessionToken] = [:]
  #endif

  public func definition() -> ModuleDefinition {
    Name("VietRidePlaces")

    AsyncFunction("beginSession") { () -> String in
      #if canImport(GooglePlaces)
      return try self.beginSessionNative()
      #else
      throw PlacesException("UNAVAILABLE", "Google Places is unavailable in this build.")
      #endif
    }

    AsyncFunction("findPredictions") { (params: [String: Any]) -> [[String: String]] in
      #if canImport(GooglePlaces)
      return try await self.findPredictionsNative(params: params)
      #else
      throw PlacesException("UNAVAILABLE", "Google Places is unavailable in this build.")
      #endif
    }

    AsyncFunction("resolvePlace") { (params: [String: Any]) -> [String: Any] in
      #if canImport(GooglePlaces)
      return try await self.resolvePlaceNative(params: params)
      #else
      throw PlacesException("UNAVAILABLE", "Google Places is unavailable in this build.")
      #endif
    }

    AsyncFunction("endSession") { (sessionId: String) in
      #if canImport(GooglePlaces)
      self.endSessionNative(sessionId: sessionId)
      #endif
    }
  }

  #if canImport(GooglePlaces)
  private func beginSessionNative() throws -> String {
    let sessionId = UUID().uuidString
    let token = GMSAutocompleteSessionToken()
    sessionLock.lock()
    sessions[sessionId] = token
    sessionLock.unlock()
    return sessionId
  }

  private func endSessionNative(sessionId: String) {
    let normalized = sessionId.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !normalized.isEmpty else { return }
    sessionLock.lock()
    sessions.removeValue(forKey: normalized)
    sessionLock.unlock()
  }

  private func token(for sessionId: String) throws -> GMSAutocompleteSessionToken {
    let normalized = sessionId.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !normalized.isEmpty else {
      throw PlacesException("INVALID_SESSION", "A Places session is required.")
    }

    sessionLock.lock()
    let token = sessions[normalized]
    sessionLock.unlock()

    guard let token else {
      throw PlacesException("INVALID_SESSION", "Places session is not active.")
    }
    return token
  }

  private func findPredictionsNative(params: [String: Any]) async throws -> [[String: String]] {
    let sessionId = stringValue(params["sessionId"])
    let query = stringValue(params["query"]).trimmingCharacters(in: .whitespacesAndNewlines)
    if query.isEmpty {
      return []
    }

    let token = try token(for: sessionId)
    let maxResults = min(max(intValue(params["maxResults"]) ?? 5, 1), 5)
    let request = GMSAutocompleteRequest(query: query)
    request.sessionToken = token

    let filter = GMSAutocompleteFilter()
    let country = stringValue(params["countryCode"]).trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    if !country.isEmpty {
      filter.countries = [country]
    }

    if let latitude = doubleValue(params["latitude"]),
       let longitude = doubleValue(params["longitude"]) {
      let radius = max(min(doubleValue(params["radiusMeters"]) ?? 5_000, 50_000), 1)
      let center = CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
      filter.locationBias = GMSPlaceCircularLocationOption(center, radius)
      filter.origin = CLLocation(latitude: latitude, longitude: longitude)
    }
    request.filter = filter

    return try await withCheckedThrowingContinuation { continuation in
      DispatchQueue.main.async {
        GMSPlacesClient.shared().fetchAutocompleteSuggestions(from: request) { results, error in
          if let error {
            continuation.resume(throwing: self.mapError(error))
            return
          }

          let payload: [[String: String]] = (results ?? []).prefix(maxResults).compactMap { suggestion in
            guard let placeSuggestion = suggestion.placeSuggestion else {
              return nil
            }
            return [
              "placeId": placeSuggestion.placeID,
              "primaryText": placeSuggestion.attributedPrimaryText?.string ?? placeSuggestion.attributedFullText.string,
              "secondaryText": placeSuggestion.attributedSecondaryText?.string ?? "",
              "fullText": placeSuggestion.attributedFullText.string,
            ]
          }
          continuation.resume(returning: payload)
        }
      }
    }
  }

  private func resolvePlaceNative(params: [String: Any]) async throws -> [String: Any] {
    let sessionId = stringValue(params["sessionId"])
    let placeId = stringValue(params["placeId"]).trimmingCharacters(in: .whitespacesAndNewlines)
    if placeId.isEmpty {
      throw PlacesException("INVALID_PLACE", "A place identifier is required.")
    }

    let token = try token(for: sessionId)
    let properties: [String] = [
      GMSPlaceProperty.placeID,
      GMSPlaceProperty.displayName,
      GMSPlaceProperty.formattedAddress,
      GMSPlaceProperty.coordinate,
    ].map(\.rawValue)

    let fetchRequest = GMSFetchPlaceRequest(
      placeID: placeId,
      placeProperties: properties,
      sessionToken: token
    )

    let place: GMSPlace = try await withCheckedThrowingContinuation { continuation in
      DispatchQueue.main.async {
        GMSPlacesClient.shared().fetchPlace(with: fetchRequest) { place, error in
          if let error {
            continuation.resume(throwing: self.mapError(error))
            return
          }
          guard let place else {
            continuation.resume(
              throwing: PlacesException("INVALID_PLACE", "Place details are unavailable.")
            )
            return
          }
          continuation.resume(returning: place)
        }
      }
    }

    let displayName = (place.displayName?.trimmingCharacters(in: .whitespacesAndNewlines)).flatMap {
      $0.isEmpty ? nil : $0
    } ?? (place.formattedAddress?.trimmingCharacters(in: .whitespacesAndNewlines)).flatMap {
      $0.isEmpty ? nil : $0
    }

    let formattedAddress = (place.formattedAddress?.trimmingCharacters(in: .whitespacesAndNewlines)).flatMap {
      $0.isEmpty ? nil : $0
    } ?? displayName

    guard let displayName, let formattedAddress else {
      throw PlacesException("INVALID_PLACE", "Place address details are incomplete.")
    }

    let coordinate = place.coordinate
    guard CLLocationCoordinate2DIsValid(coordinate) else {
      throw PlacesException("INVALID_PLACE", "Place coordinates are unavailable.")
    }

    endSessionNative(sessionId: sessionId)

    return [
      "placeId": placeId,
      "displayName": displayName,
      "formattedAddress": formattedAddress,
      "latitude": coordinate.latitude,
      "longitude": coordinate.longitude,
    ]
  }

  private func mapError(_ error: Error) -> PlacesException {
    let nsError = error as NSError
    let message = nsError.localizedDescription
    let isPlacesError = nsError.domain == kGMSPlacesErrorDomain
    let configurationCodes: Set<Int> = [-4, -5, -9, -10]
    let quotaCodes: Set<Int> = [-6, -7, -8]

    if (isPlacesError && configurationCodes.contains(nsError.code))
      || message.localizedCaseInsensitiveContains("API key")
      || message.localizedCaseInsensitiveContains("not authorized")
      || message.localizedCaseInsensitiveContains("REQUEST_DENIED") {
      return PlacesException("CONFIGURATION", "Google Places is not configured for this build.")
    }
    if (isPlacesError && quotaCodes.contains(nsError.code))
      || message.localizedCaseInsensitiveContains("OVER_QUERY_LIMIT")
      || message.localizedCaseInsensitiveContains("quota")
      || message.localizedCaseInsensitiveContains("rate limit") {
      return PlacesException("QUOTA", "Places quota has been exceeded.")
    }
    if (isPlacesError && nsError.code == -1)
      || message.localizedCaseInsensitiveContains("network")
      || message.localizedCaseInsensitiveContains("offline")
      || message.localizedCaseInsensitiveContains("Internet")
      || nsError.domain == NSURLErrorDomain {
      return PlacesException("OFFLINE", "Places requires a network connection.")
    }
    return PlacesException("UNAVAILABLE", "Places request failed.")
  }

  private func stringValue(_ value: Any?) -> String {
    if let value = value as? String {
      return value
    }
    if let value = value as? NSString {
      return value as String
    }
    return ""
  }

  private func doubleValue(_ value: Any?) -> Double? {
    if let value = value as? Double {
      return value
    }
    if let value = value as? NSNumber {
      return value.doubleValue
    }
    return nil
  }

  private func intValue(_ value: Any?) -> Int? {
    if let value = value as? Int {
      return value
    }
    if let value = value as? NSNumber {
      return value.intValue
    }
    return nil
  }
  #endif
}

private final class PlacesException: Exception {
  private let codeValue: String
  private let messageValue: String

  init(_ code: String, _ message: String) {
    self.codeValue = code
    self.messageValue = message
    super.init()
  }

  override var code: String {
    codeValue
  }

  override var reason: String {
    messageValue
  }
}

File: modules/vietride-places/android/src/main/java/com/vietride/places/VietRidePlacesModule.kt
package com.vietride.places

import android.content.pm.PackageManager
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.tasks.CancellationTokenSource
import com.google.android.libraries.places.api.Places
import com.google.android.libraries.places.api.model.AutocompleteSessionToken
import com.google.android.libraries.places.api.model.CircularBounds
import com.google.android.libraries.places.api.model.Place
import com.google.android.libraries.places.api.net.FetchPlaceRequest
import com.google.android.libraries.places.api.net.FindAutocompletePredictionsRequest
import com.google.android.libraries.places.api.net.PlacesClient
import com.google.android.libraries.places.api.net.PlacesStatusCodes
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

class FindPredictionsArgs : Record {
  @Field
  var sessionId: String = ""

  @Field
  var query: String = ""

  @Field
  var latitude: Double? = null

  @Field
  var longitude: Double? = null

  @Field
  var radiusMeters: Double? = null

  @Field
  var countryCode: String? = null

  @Field
  var maxResults: Int? = null
}

class ResolvePlaceArgs : Record {
  @Field
  var sessionId: String = ""

  @Field
  var placeId: String = ""

  /**
   * When true (default), autocomplete session is closed after a successful details fetch.
   * Map-pin previews should pass false (and may omit sessionId) so multiple pins can load.
   */
  @Field
  var endSession: Boolean = true
}

class VietRidePlacesModule : Module() {
  private val sessions = ConcurrentHashMap<String, AutocompleteSessionToken>()
  private val activeCancellations = ConcurrentHashMap<String, CancellationTokenSource>()
  private var placesClient: PlacesClient? = null

  override fun definition() = ModuleDefinition {
    Name("VietRidePlaces")

    OnDestroy {
      activeCancellations.values.forEach { it.cancel() }
      activeCancellations.clear()
      sessions.clear()
      placesClient = null
    }

    AsyncFunction("beginSession") { promise: Promise ->
      try {
        ensurePlacesInitialized()
        val sessionId = UUID.randomUUID().toString()
        sessions[sessionId] = AutocompleteSessionToken.newInstance()
        promise.resolve(sessionId)
      } catch (error: Exception) {
        promise.reject(mapException(error))
      }
    }

    AsyncFunction("findPredictions") { args: FindPredictionsArgs, promise: Promise ->
      try {
        val client = ensurePlacesInitialized()
        val sessionId = args.sessionId.trim()
        val query = args.query.trim()
        if (sessionId.isEmpty()) {
          throw PlacesCodedException("INVALID_SESSION", "A Places session is required.")
        }
        if (query.isEmpty()) {
          promise.resolve(emptyList<Map<String, String>>())
          return@AsyncFunction
        }

        val token = sessions[sessionId]
          ?: throw PlacesCodedException("INVALID_SESSION", "Places session is not active.")

        activeCancellations.remove(sessionId)?.cancel()
        val cancellation = CancellationTokenSource()
        activeCancellations[sessionId] = cancellation

        val builder = FindAutocompletePredictionsRequest.builder()
          .setQuery(query)
          .setSessionToken(token)
          .setCancellationToken(cancellation.token)

        val latitude = args.latitude
        val longitude = args.longitude
        if (latitude != null && longitude != null) {
          val center = LatLng(latitude, longitude)
          val radius = (args.radiusMeters ?: DEFAULT_BIAS_RADIUS_METERS).coerceIn(1.0, 50_000.0)
          builder.setLocationBias(CircularBounds.newInstance(center, radius))
          builder.setOrigin(center)
        }

        val country = args.countryCode?.trim()?.lowercase()
        if (!country.isNullOrEmpty()) {
          builder.setCountries(listOf(country))
        }

        client.findAutocompletePredictions(builder.build())
          .addOnSuccessListener { response ->
            if (cancellation.token.isCancellationRequested) {
              // Every Expo Promise must settle. JS request sequencing ignores
              // this stale empty result after a newer query starts.
              promise.resolve(emptyList<Map<String, String>>())
              return@addOnSuccessListener
            }
            val limit = (args.maxResults ?: DEFAULT_MAX_RESULTS).coerceIn(1, 5)
            val payload = response.autocompletePredictions
              .take(limit)
              .map { prediction ->
                mapOf(
                  "placeId" to prediction.placeId,
                  "primaryText" to prediction.getPrimaryText(null).toString(),
                  "secondaryText" to prediction.getSecondaryText(null).toString(),
                  "fullText" to prediction.getFullText(null).toString(),
                )
              }
            promise.resolve(payload)
          }
          .addOnFailureListener { error ->
            if (cancellation.token.isCancellationRequested) {
              promise.resolve(emptyList<Map<String, String>>())
              return@addOnFailureListener
            }
            promise.reject(mapException(error))
          }
          .addOnCompleteListener {
            activeCancellations.remove(sessionId, cancellation)
          }
      } catch (error: Exception) {
        promise.reject(mapException(error))
      }
    }

    AsyncFunction("resolvePlace") { args: ResolvePlaceArgs, promise: Promise ->
      try {
        val client = ensurePlacesInitialized()
        val sessionId = args.sessionId.trim()
        val placeId = args.placeId.trim()
        if (placeId.isEmpty()) {
          throw PlacesCodedException("INVALID_PLACE", "A place identifier is required.")
        }

        // Session is optional: map pin previews resolve without a session so multiple
        // markers can load without closing the autocomplete session early.
        val token = if (sessionId.isEmpty()) {
          null
        } else {
          sessions[sessionId]
            ?: throw PlacesCodedException("INVALID_SESSION", "Places session is not active.")
        }

        val requestBuilder = FetchPlaceRequest.builder(placeId, PLACE_FIELDS)
        if (token != null) {
          requestBuilder.setSessionToken(token)
        }
        val request = requestBuilder.build()

        client.fetchPlace(request)
          .addOnSuccessListener { response ->
            val place = response.place
            val location = place.location
            if (location == null) {
              promise.reject(
                PlacesCodedException(
                  "INVALID_PLACE",
                  "Place coordinates are unavailable for id=$placeId",
                ),
              )
              return@addOnSuccessListener
            }
            // Prefer Places fields; never invent a name when both are blank.
            val displayName = place.displayName?.takeIf { it.isNotBlank() }
              ?: place.formattedAddress?.takeIf { it.isNotBlank() }
              ?: ""
            val formattedAddress = place.formattedAddress?.takeIf { it.isNotBlank() }
              ?: displayName
            if (displayName.isBlank() || formattedAddress.isBlank()) {
              promise.reject(
                PlacesCodedException(
                  "INVALID_PLACE",
                  "Place address details are incomplete for id=$placeId",
                ),
              )
              return@addOnSuccessListener
            }

            if (args.endSession && sessionId.isNotEmpty()) {
              sessions.remove(sessionId)
              activeCancellations.remove(sessionId)?.cancel()
            }

            // Box as Double so the JS bridge always receives finite numbers.
            promise.resolve(
              mapOf(
                "placeId" to placeId,
                "displayName" to displayName,
                "formattedAddress" to formattedAddress,
                "latitude" to location.latitude.toDouble(),
                "longitude" to location.longitude.toDouble(),
              ),
            )
          }
          .addOnFailureListener { error ->
            promise.reject(mapException(error))
          }
      } catch (error: Exception) {
        promise.reject(mapException(error))
      }
    }

    AsyncFunction("endSession") { sessionId: String, promise: Promise ->
      val normalized = sessionId.trim()
      if (normalized.isNotEmpty()) {
        sessions.remove(normalized)
        activeCancellations.remove(normalized)?.cancel()
      }
      promise.resolve(null)
    }
  }

  @Synchronized
  private fun ensurePlacesInitialized(): PlacesClient {
    placesClient?.let { return it }

    val context = appContext.reactContext
      ?: throw PlacesCodedException("UNAVAILABLE", "React context is not ready.")

    if (!Places.isInitialized()) {
      val apiKey = readNativeMapsApiKey(context)
        ?: throw PlacesCodedException(
          "CONFIGURATION",
          "Google Places is not configured for this build.",
        )
      try {
        Places.addInternalUsageAttributionId(GOOGLE_MAPS_USAGE_ATTRIBUTION_ID)
        Places.initializeWithNewPlacesApiEnabled(context.applicationContext, apiKey)
      } catch (error: Exception) {
        throw PlacesCodedException(
          "CONFIGURATION",
          "Google Places could not be initialized.",
          error,
        )
      }
    }

    val client = Places.createClient(context.applicationContext)
    placesClient = client
    return client
  }

  private fun readNativeMapsApiKey(context: android.content.Context): String? {
    return try {
      val applicationInfo = context.packageManager.getApplicationInfo(
        context.packageName,
        PackageManager.GET_META_DATA,
      )
      val raw = applicationInfo.metaData?.getString(ANDROID_MAPS_API_KEY_META)
        ?.trim()
        .orEmpty()
      if (
        raw.isEmpty()
        || raw.contains("YOUR_KEY", ignoreCase = true)
        || raw.contains("PLACEHOLDER", ignoreCase = true)
        || raw.startsWith("TEST_", ignoreCase = true)
      ) {
        null
      } else {
        raw
      }
    } catch (_: Exception) {
      null
    }
  }

  private fun mapException(error: Throwable): CodedException {
    if (error is PlacesCodedException) {
      return error
    }

    if (error is ApiException) {
      return when (error.statusCode) {
        7 -> PlacesCodedException("OFFLINE", "Places requires a network connection.")
        8 -> PlacesCodedException("UNAVAILABLE", "Places is temporarily unavailable.")
        13 -> PlacesCodedException("UNAVAILABLE", "Places is temporarily unavailable.")
        STATUS_CANCELLED -> PlacesCodedException("UNAVAILABLE", "Places request was cancelled.")
        PlacesStatusCodes.OVER_QUERY_LIMIT -> PlacesCodedException("QUOTA", "Places quota has been exceeded.")
        PlacesStatusCodes.REQUEST_DENIED -> PlacesCodedException(
          "CONFIGURATION",
          "Google Places is not configured for this build.",
        )
        // Common when a Maps POI placeId is not resolvable via Place Details.
        PlacesStatusCodes.NOT_FOUND -> PlacesCodedException(
          "INVALID_PLACE",
          "Place was not found (status ${error.statusCode}).",
        )
        PlacesStatusCodes.INVALID_REQUEST -> PlacesCodedException(
          "INVALID_PLACE",
          "Invalid Google Places request (status ${error.statusCode}).",
        )
        else -> PlacesCodedException(
          "UNAVAILABLE",
          "Places request failed (status ${error.statusCode}).",
        )
      }
    }

    val message = error.message.orEmpty()
    return when {
      message.contains("API_KEY", ignoreCase = true)
        || message.contains("not authorized", ignoreCase = true)
        || message.contains("REQUEST_DENIED", ignoreCase = true) ->
        PlacesCodedException("CONFIGURATION", "Google Places is not configured for this build.")
      message.contains("OVER_QUERY_LIMIT", ignoreCase = true)
        || message.contains("RESOURCE_EXHAUSTED", ignoreCase = true) ->
        PlacesCodedException("QUOTA", "Places quota has been exceeded.")
      message.contains("Unable to resolve host", ignoreCase = true)
        || message.contains("network", ignoreCase = true)
        || message.contains("timeout", ignoreCase = true) ->
        PlacesCodedException("OFFLINE", "Places requires a network connection.")
      else -> PlacesCodedException("UNAVAILABLE", "Places request failed.")
    }
  }

  companion object {
    private const val ANDROID_MAPS_API_KEY_META = "com.google.android.geo.API_KEY"
    private const val GOOGLE_MAPS_USAGE_ATTRIBUTION_ID = "gmp_git_agentskills_v1"
    private const val STATUS_CANCELLED = 16
    private const val DEFAULT_BIAS_RADIUS_METERS = 5_000.0
    private const val DEFAULT_MAX_RESULTS = 5
    private val PLACE_FIELDS = listOf(
      Place.Field.ID,
      Place.Field.DISPLAY_NAME,
      Place.Field.FORMATTED_ADDRESS,
      Place.Field.LOCATION,
    )
  }
}

private class PlacesCodedException(
  code: String,
  message: String,
  cause: Throwable? = null,
) : CodedException(code, message, cause)
